app/src/main/java/org/p2p/wallet/striga/offramp/ui/StrigaOffRampPresenter.kt

```diff
@@ -0,0 +1,344 @@
+        get() = isErrorHappened
+
+    override fun attach(view: View) {
+        super.attach(view)
+        interactor.startPolling(this)
+        observeUsdc()
+        observeRateChanges()
+        observeInternetState()
+        observeViewState()
+    }
+
+    override fun detach() {
+        super.detach()
+        interactor.stopPolling()
+    }
+
+    private fun observeUsdc() {
+        launch {
+            tokenServiceCoordinator.observeUserTokens()
+                .collect { handleTokenState(it) }
+        }
+    }
+
+    private fun observeRateChanges() {
+        launch {
+            interactor.observeExchangeRateState()
+                .collect { state -> handleRateChange(state) }
+        }
+    }
+
+    private fun observeInternetState() {
+        launch {
+            connectionManager.connectionStatus.collect { hasConnection ->
+                if (hasConnection) {
+                    interactor.startPolling(this)
+
+                    // restore after error
+                    if (isErrorHappened) {
+                        isErrorHappened = false
+                        tokenServiceCoordinator.refresh()
+                    }
+                } else {
+                    isErrorHappened = true
+                    interactor.stopPolling()
+
+                    setExchangeRateState(strigaOffRampMapper.mapRateError())
+                    setTokensAreDisabled()
+                    setButtonState(StrigaOffRampButtonState.ErrorGeneral)
+
+                    view?.showUiKitSnackBar(messageResId = R.string.error_no_internet_message)
+                }
+            }
+        }
+    }
+
+    private fun observeViewState() {
+        launch {
+            viewState.collect {
+                view?.setRatioState(rateTickerMapper.mapRate(it.exchangeRateState))
+                view?.setButtonState(it.buttonState)
+            }
+        }
+    }
+
+    override fun onTokenAAmountChange(amountA: String) {
+        inputAmountA = amountA.toBigDecimalOrZero()
+        setAndCalcTokenAmount(
+            targetTokenType = StrigaOffRampTokenType.TokenB,
+            amount = inputAmountA
+        )
+        validate()
+    }
+
+    override fun onTokenBAmountChange(amountB: String) {
+        inputAmountB = amountB.toBigDecimalOrZero()
+        setAndCalcTokenAmount(
+            targetTokenType = StrigaOffRampTokenType.TokenA,
+            amount = inputAmountB
+        )
+        validate()
+    }
+
+    override fun onAllAmountClick() {
+        // ignore all amount click when no internet or rate is not loaded/error occurred
+        if (!connectionManager.connectionStatus.value || cannotClickAllAmount) return
+
+        setTokenAmount(
+            targetTokenType = StrigaOffRampTokenType.TokenA,
+            amount = balance
+        )
+        setAndCalcTokenAmount(
+            targetTokenType = StrigaOffRampTokenType.TokenB,
+            amount = balance
+        )
+        validate()
```

Comment:
setAndCalcTokenAmount always comes with `validate`, maybe put it there?

Reply:
Yes and no. Moving `validate` call to the setTokenAmount breaks the button behavior when an error occurred and/or the internet is restored; that "restoring" logic have their own `setButtonState` calls with a specific button state for only that cases.
Let's keep the `validate` calls as is, explicit, for a while.